test/extended/dr/machine_recover.go

```diff
@@ -0,0 +1,388 @@
+				o.Expect(err).NotTo(o.HaveOccurred())
+
+				// add controller reference
+				replacedMasterMachineCopy := replacedMasterMachine.DeepCopy()
+				ownerReferences := getOwnerReferenceForMasterMachine(replacedMasterMachineCopy)
+				replacedMasterMachineCopy.SetOwnerReferences(ownerReferences)
+				_, err = ms.Update(context.Background(), replacedMasterMachineCopy, metav1.UpdateOptions{})
+				o.Expect(err).NotTo(o.HaveOccurred())
+
+				targets := []*corev1.Node{replacedMaster, replacedWorker}
+				targetMachineNames := []string{replacedMasterMachineName, replacedWorkerMachineName}
+
+				// we use a hard shutdown to simulate a poweroff
+				framework.Logf("Forcing shutdown of node %s", replacedMaster.Name)
+				_, err = ssh("sudo -i systemctl poweroff --force --force", replacedMaster)
+				framework.Logf("Forcing shutdown of node %s", replacedWorker.Name)
+				_, err = ssh("sudo -i systemctl poweroff --force --force", replacedWorker)
+
+				pollConfig := rest.CopyConfig(config)
+				pollConfig.Timeout = 5 * time.Second
+				pollClient, err := kubernetes.NewForConfig(pollConfig)
+				o.Expect(err).NotTo(o.HaveOccurred())
+
+				framework.Logf("Wait for nodes to go unready")
+				err = wait.Poll(30*time.Second, 30*time.Minute, func() (done bool, err error) {
+					nodes, err := pollClient.CoreV1().Nodes().List(context.Background(), metav1.ListOptions{})
+					if err != nil || nodes.Items == nil {
+						framework.Logf("return false - err %v nodes.Items %v", err, nodes.Items)
+						return false, nil
+					}
+					notReadyNodes := sets.NewString()
+					for _, node := range nodes.Items {
+						for _, t := range node.Spec.Taints {
+							if t.Key == "node.kubernetes.io/unreachable" {
+								notReadyNodes.Insert(node.Name)
+								continue
+							}
+						}
+					}
+					if notReadyNodes.Len() != len(targets) {
+						framework.Logf("Nodes waiting to go unready: %v", notReadyNodes.List())
+						return false, nil
+					}
+					return true, nil
+				})
+				o.Expect(err).NotTo(o.HaveOccurred())
+
+				framework.Logf("Wait for machines to be deleted")
+				err = wait.Poll(30*time.Second, 30*time.Minute, func() (done bool, err error) {
+					machines, err := ms.List(context.Background(), metav1.ListOptions{})
+					if err != nil || machines.Items == nil {
+						framework.Logf("return false - err %v nodes.Items %v", err, machines.Items)
+						return false, nil
+					}
+					vanishedMachines := sets.NewString()
+					for _, machine := range targetMachineNames {
+						vanishedMachines.Insert(machine)
+					}
+					for _, machine := range machines.Items {
+						vanishedMachines.Delete(machine.GetName())
+					}
+					if vanishedMachines.Len() != len(targetMachineNames) {
+						framework.Logf("Machines waiting to go be deleted: %v", vanishedMachines.List())
+						return false, nil
+					}
+					return true, nil
+				})
+				o.Expect(err).NotTo(o.HaveOccurred())
+
```

Comment:
at this point, we need to rsh into one of the remaining etcd pods and remove the member. CEO does not currently have an automated way of scaling down. We removed the feature in 4.4 to reduce the scope during rework.